### ISSUE TITLE
Discrepancy in the Eval Accuracy of Expert SyncNet Discriminator

### ISSUE THREAD

**Issue by amos-ahtrahddis** [Rudrabha/Wav2Lip#363]
1. It is mentioned in the paper that the expert lip-sync discriminator used for training the Wav2Lip model is about **91%** accurate on the LRS2 test set. However, when evaluated on the LRS2 test set, the accuracy obtained is around **80%** which is much lower than the value mentioned in the paper.

Below is the code used for computing the accuracy. This function returns the number of correct predictions for a given batch.

```
def get_corr_count(a, v, y):

    loss, dist = cosine_loss(a, v, y, return_dist=True) # returning distances from cosine_loss
    y = y.squeeze()
    preds = (dist >= 0.5)
    corr_count = torch.sum(preds == y)
    return corr_count, len(y)
```

2. With a model trained using the released code and configs, we were able to obtain an eval accuracy of around **88%** on the LRS2 test dataset.

Can anybody comment on the discrepancies in the eval performance of their pre-trained model and whether you were able to obtain the mentioned eval accuracy of 91%?

**Comment by Mayur28** [comment]
Hi @amos-ahtrahddis ,

I just wanted to get some clarity on your question please. In your first point, you state that when you evaluated on the LRS2 test dataset, you achieved an accuracy of 80% - Am I correct in saying that this is when you used the pre-trained model and the 88% that you mention in point 2 is achieved when you trained the model from scratch (using the code in the repo)?

Thanks

**Comment by amos-ahtrahddis** [comment]
Hi @Mayur28 

Yup, you're right! Using their pre-trained, the eval accuracy obtained was around 80% and with the model trained using their code I was able to achieve an eval accuracy of around 88%.

**Comment by Mayur28** [comment]
Thanks for clarifying @amos-ahtrahddis.

**Comment by amos-ahtrahddis** [comment]
Hi @Mayur28 !
Were you able to verify their claimed eval accuracy and reproduce their results? 
Can you comment on these discrepancies or other observations you have made regarding this?

**Comment by Mayur28** [comment]
Hi @amos-ahtrahddis,

I've managed to evaluate the "expert" lip-sync discriminator, I am also only able to achieve an accuracy of 81.882%. To give more insight into how I evaluated the model, I created 10 SyncNet evaluation filelists that stores the following info:
<Video Name> <Start frame (for audio)> <stat frame (for video)>
These "filelists" are generated randomly, and because of the way I have it, the exact same in-sync and out-of-sync pairs are used for evaluating all my models. For each "filelist" I calculate the accuracy, and I then take the average of these 10 accuracies.
I'm yet to train the the Wav2Lip model from scratch and evaluate, but once I do so, I will inform you.

**Comment by amos-ahtrahddis** [comment]
Hi @Mayur28! Sorry for the late reply, have been working on few other projects lately. 
Thank you so much for trying this out! Seems like the results obtained by you are also in the same ball park.

Sure, do let me know how your models turn out. Curious to see your results!

**Comment by hannarud** [comment]
Hi @amos-ahtrahddis @Mayur28!

I noticed a bit unexpected behavior for syncnet model, but maybe it can clarify something.
If I change this row https://github.com/Rudrabha/Wav2Lip/blob/b9759a3467cb1b7519f1a3b91f5a84cb4bc1ae4a/color_syncnet_train.py#L82 by either leaving here only `True` (so that the Dataloader always returns in-sync pairs), only `False` (so that the Dataloader always returns out-of-sync pairs) and leaving this row as-is (returning in-sync and out-of-sync pairs nearly half of the time), I obtain pretty different results:
* when leaving this line as-is (half of samples in-sync, half out-of-sync) I get the loss value around `0.8` each time I evaluate, and if speaking of accuracy (computed with the function @amos-ahtrahddis provided in his first comment) accuracy rate is `75%` - I guess those are the values you and @Mayur28 obtained. So I confirm them.
* when setting `True` values only in this line (i.e. returning only in-sync samples every time), I get the loss value around `1.4`! Which is huge, in my opinion. Accuracy here is only `60%`.
* when I set Dataloader to return only out-of-sync samples (only `False` in this line), I get the results claimed by the authors in the article and other issues (https://github.com/Rudrabha/Wav2Lip/issues/111, for example): the loss value is around `0.2-0.24`, accuracy rate is `90-91%`

I'm not using LRS for testing, but I believe my statement generalizes on different datasets, because I tested different samples many times.

What I really don't understand is that whenever I train my own syncnet model, loss and accuracy values are rather stable with any True-False combinations I pass for evaluation. Like, if I get `0.3` loss, it will be the same for any of these options, and accuracy will be around `89%`. What is the trick? :) @prajwalkr why does original syncnet checkpoint have this non-uniform behavior?
